fix SDK6 heap dumper documentation to reflect actual usage. (#884)
MicroEJ/docs: SDK6UserGuide/heapAnalyzer.rst

Heap Dumper
-----------

The Heap Dumper generates heap files either :
- directly from the Java heap on simulator 

> - from .hex files that must be manually retrieved on device.

Reviewer: ```suggestion
- from .hex files that must be manually retrieved from the device.
```

Author: applied in https://github.com/MicroEJ/docs/pull/884/commits/437e9770e60d925051a25d490677e7d73069f934

> The heap dump should be performed after a call to ``System.gc()`` to exclude discardable objects.

Reviewer: the dump can be generation in many cases not only after a `System.gc` see here https://github.com/MicroEJ/docs/blob/master/ApplicationDeveloperGuide/debuggerProxy.rst?plain=1#L23

Author: Explanation added in https://github.com/MicroEJ/docs/pull/884/commits/9272f8c920fe424effb96bb4b13a4af7ae5bc4ec

When the `System.gc()`_ method is called, 
a ``.heap`` file is generated in the ``build/output/application/heapDump/`` folder of the Application project.

Device
^^^^^^

In order to generate a Heap dump of an Application running on a device:

- Update your Application code to call the `System.gc()`_ method where you need a Heap dump.
- Build the Executable and deploy it on the device.
- Start a debug session.
- Add a breakpoint to ``LLMJVM_on_Runtime_gc_done`` Core Engine hook. This function is called by the Core Engine when `System.gc()`_ method is done.
  Alternatively, if you are experiencing out of memory errors, you can directly add a breakpoint to the ``LLMJVM_on_OutOfMemoryError_thrown`` Core Engine hook.
- Resume the execution until the breakpoint is reached. You are now ready to dump the memory. Next steps are:
  
  - :ref:`Retrieve the hex file from the device <sdk6_heapdumper_get_hex>`
  - :ref:`Extract the Heap dump from the hex file <sdk6_heapdumper_extract_heap>`

.. note::

   Core Engine hooks may have been inlined by the third-party linker. 
   If the symbol is not accessible to the debugger, you must declare them in your VEE Port:

   .. code:: c

>       void LLMJVM_on_Runtime_gc_done(){

Reviewer 2: confirmed by @tboue

         //No need to add code to the function
      }

      void LLMJVM_on_OutOfMemoryError_thrown(){
         //No need to add code to the function

>       }

Author: Tested again, no need to implement the callback, just having a body header is enough

- Update your Application code to call the ``System.gc()`` method where you need a Heap dump.
- Build the Executable and deploy it on the device.

When the Application is executed on the device, the method ``System.gc()`` ends by calling ``LLMJVM_on_Runtime_gc_done`` callback.
This method isn't always implemented as it is mostly used to perform a heap dump right after the garbage collection.
If the symbol isn't embedded, add the following lines in any source file (ex:main.c):

.. code:: c

>    void LLMJVM_on_OutOfMemoryError_thrown(){

Reviewer: this is the wrong example, you should define here the LLMJVM_on_Runtime_gc_done

Author: fixed in https://github.com/MicroEJ/docs/pull/884/commits/9272f8c920fe424effb96bb4b13a4af7ae5bc4ec

- Update your Application code to call the ``System.gc()`` method where you need a Heap dump.
- Build the Executable and deploy it on the device.

When the Application is executed on the device, the method ``System.gc()`` ends by calling ``LLMJVM_on_Runtime_gc_done`` callback.
This method isn't always implemented as it is mostly used to perform a heap dump right after the garbage collection.

> If the symbol isn't embedded, add the following lines in any source file (ex:main.c):

Reviewer: I don't think this is right, I purpose something like this
```suggestion
If the symbol is inlined, add the following lines in any source file (ex:main.c):
```
What do you think @frivieremicroej?

Author: Changed to https://github.com/MicroEJ/docs/pull/884/files#diff-d122d22cbbd8db54cff58c20f6831435524140fb2fafe25be7c8dcc084801a51R113-R114

You now have the ``.hex`` file and need to extract the Heap dump.

When using the tool in a multi-application context, additionnal sections must be dumped:
- The dynamic features table:

>    dump ihex memory java_features_dynamic_start java_features_dynamic_end

Reviewer: Is this correct? or
```suggestion
   dump ihex memory &java_features_dynamic_start &java_features_dynamic_end
```

> - features related sections. These will change depending on the board, number of features and other parameters. It is possible to dump the entire memory and let the :ref:`heap extractor<_sdk6_heapdumper_extract_heap>` sort the requiered sections instead.

Reviewer: ```suggestion
- features related sections. These will change depending on the board, number of features and other parameters. It is possible to dump the entire memory and let the :ref:`heap extractor<_sdk6_heapdumper_extract_heap>` extract the requiered sections instead.
```
